src/ConnectionPool.jl

```diff
@@ -0,0 +1,575 @@
+    return
+end
+
+
+"""
+    purge(::Connection)
+
+Remove unread data from a `Connection`.
+"""
+
+function purge(c::Connection)
+    @require !isopen(c.io)
+    while !eof(c.io)
+        readavailable(c.io)
+    end
+    c.excess = nobytes
+    @ensure nb_available(c) == 0
+end
+
+
+"""
+The `pool` is a collection of open `Connection`s.  The `request`
+function calls `getconnection` to retrieve a connection from the
+`pool`.  When the `request` function has written a Request Message
+it calls `closewrite` to signal that the `Connection` can be reused
+for writing (to send the next Request). When the `request` function
+has read the Response Message it calls `closeread` to signal that
+the `Connection` can be reused for reading.
+"""
+
+const pool = Vector{Connection}()
+const poollock = ReentrantLock()
+const poolcondition = Condition()
+
+"""
+    closeall()
+
+Close all connections in `pool`.
+"""
+
+function closeall()
+
+    lock(poollock)
+    for c in pool
+        close(c)
+    end
+    empty!(pool)
+    unlock(poollock)
+    notify(poolcondition)
+    return
+end
+
+
+"""
+    findwritable(type, host, port) -> Vector{Connection}
+
+Find `Connections` in the `pool` that are ready for writing.
+"""
+
+function findwritable(T::Type,
+                      host::AbstractString,
+                      port::AbstractString,
+                      pipeline_limit::Int,
+                      reuse_limit::Int)
+
+    filter(c->(!c.writebusy &&
+               typeof(c.io) == T &&
+               c.host == host &&
+               c.port == port &&
+               c.pipeline_limit == pipeline_limit &&
+               c.writecount < reuse_limit &&
+               c.writecount - c.readcount < pipeline_limit + 1 &&
+               isopen(c.io)), pool)
+end
+
+
+"""
+    findoverused(type, host, port, reuse_limit) -> Vector{Connection}
+
+Find `Connections` in the `pool` that are over the reuse limit
+and have no more active readers.
+"""
+
+function findoverused(T::Type,
+                      host::AbstractString,
+                      port::AbstractString,
+                      reuse_limit::Int)
+
+    filter(c->(typeof(c.io) == T &&
+               c.host == host &&
+               c.port == port &&
+               c.readcount >= reuse_limit &&
+               !c.readbusy &&
+               isopen(c.io)), pool)
+end
+
+
+"""
+    findall(type, host, port) -> Vector{Connection}
+
+Find all `Connections` in the `pool` for `host` and `port`.
+"""
+
+function findall(T::Type,
+                 host::AbstractString,
+                 port::AbstractString,
+                 pipeline_limit::Int)
+
+    filter(c->(typeof(c.io) == T &&
+               c.host == host &&
+               c.port == port &&
+               c.pipeline_limit == pipeline_limit &&
+               isopen(c.io)), pool)
+end
+
+
+"""
+    purge()
+
+Remove closed connections from `pool`.
+"""
+function purge()
+    while (i = findfirst(x->!isopen(x.io), pool)) > 0
```

Comment:
Personally I find this hard to tell what's going on; it seems to be doing Onlogn work when a single pass over the connection pool vector should suffice (accumulating indices of connections to delete and deleting them all at once after)

Reply:
I think the work is irrelevant in this case, if there is one delete per 100 requests on a connection,  this function will not be a significant factor.
But, looking at `?deleteat!` I see it has an option for passing a `Vector{Bool}`. Nice! So we could write this:
```julia
 deleteat!(pool, map(x->!isopen(x.io), pool))
```
(... with some extra stuff to maintain the debug output.)

Reply:
```julia
function purge()
    isdeleteable(c) = !isopen(x.io) && (@debug 1 "🗑  Deleted:        $c"; true)
    deleteat!(pool, map(isdeletable, pool))
end
```
?

Reply:
Looks great

Reply:
cb9ad328c7348d811147c1bd511a5b5be3b0fb04